resources/net.go

```diff
@@ -0,0 +1,849 @@
+func (obj *NetRes) Init() error {
+	var err error
+	// tmp directory for pipe socket
+	// XXX: Replace with obj.VarDir
+	if obj.tempDir, err = ioutil.TempDir("", "pipe"); err != nil {
+		return errwrap.Wrapf(err, "could not get TempDir")
+	}
+
+	// store the network interface in the struct
+	obj.iface = &iface{}
+	if obj.iface.Interface, err = net.InterfaceByName(obj.GetName()); err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+	// store the netlink link to use as interface input in netlink functions
+	if obj.iface.link, err = netlink.LinkByName(obj.GetName()); err != nil {
+		return errwrap.Wrapf(err, "error finding link: %s", obj.GetName())
+	}
+
+	// build the path to the networkd configuration file
+	obj.unitFilePath = networkdUnitFileDir + IfacePrefix + obj.GetName() + networkdUnitFileExt
+
+	return obj.BaseRes.Init()
+}
+
+// Watch listens for events from the specified interface via a netlink socket.
+// XXX: currently gets events from ALL interfaces, would be nice to reject
+// events from other interfaces.
+func (obj *NetRes) Watch() error {
+	// waitgroup for netlink receive goroutine
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+
+	// channel for netlink messages
+	nlChan := make(chan *nlChanStruct) // closed from goroutine
+
+	// create a pipe socket to unblock unix.Select when we close
+	pipe, err := unix.Socket(unix.AF_UNIX, unix.SOCK_RAW, unix.PROT_NONE)
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating pipe socket")
+	}
+	// bind the pipe to a file
+	if err = unix.Bind(pipe, &unix.SockaddrUnix{
+		Name: path.Join(obj.tempDir, "pipe.sock"),
+	}); err != nil {
+		return errwrap.Wrapf(err, "error binding pipe socket")
+	} // closed when the goroutine returns to avoid race with sendto below
+
+	// unblock select if it didn't catch the interrupt
+	defer unix.Sendto(pipe, nil, 0, &unix.SockaddrUnix{
+		Name: path.Join(obj.tempDir, "pipe.sock"),
+	})
+
+	// make a netlink socket file descriptor
+	nlSock, err := unix.Socket(unix.AF_NETLINK, unix.SOCK_RAW, unix.NETLINK_ROUTE)
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating netlink socket")
+	}
+	// bind to the socket and add add the netlink groups we need to get events
+	if err := unix.Bind(nlSock, &unix.SockaddrNetlink{
+		Family: unix.AF_NETLINK,
+		Groups: rtmGrpLink | rtmGrpIPv4IfAddr | rtmGrpIPv4IfRoute | rtmGrpIPv6IfAddr, // 0x1 | 0x10 | 0x40 | 0x100
+	}); err != nil {
+		return errwrap.Wrapf(err, "error binding netlink socket")
+	}
+	defer unix.Close(nlSock)
+	defer unix.Shutdown(nlSock, unix.SHUT_RDWR) // probably unnecessary, but can't hurt
+
+	// watch the systemd-networkd configuration file
+	recWatcher, err := recwatch.NewRecWatcher(obj.unitFilePath, false)
+	if err != nil {
+		return err
+	}
+	// close the recwatcher when we're done
+	defer recWatcher.Close()
+
+	// channel to unblock selects in goroutine
+	closeChan := make(chan struct{})
+	defer close(closeChan)
+
+	wg.Add(1)
```

Comment:
Since the pipe watching code is sufficiently complicated, we might want to refactor it out into a separate self contained struct (with associated methods) so that:

1) This code becomes more readable.
2) It's a standalone piece that we could use elsewhere, in particular in a test :) [future patch]

So backup this code first, but then build a small API for this, which we'll then consume here. You can include this all at the bottom of the file.